### DISCUSSION TITLE
[Question about order of calculation]

### DISCUSSION THREAD

**Discussion by coskrrb2002** [idaholab/moose#17543] [Q&A General] [answered]
Dear moose users,

I have a question about calculation steps of auxkernel and kernel.

In my numerical model, I am trying to use an auxkernel to calculate a value, and this value is used in kernels to find a solution for numerical model.

But I am not sure that auxkernel is calculated first, and it used for the calculation in kernels.

So simply my question :
Auxkernel is calculated before kernel to calculate residual of variables? (1)
After Kernel calculation? (2)
Simulataneously? (3)

Calculation steps:
1) Auxkernel -> Kernel -> Find solution
2) Kernel -> Auxkernel -> Find solution
3) Kernel <=> Auxkernel -> Find solution

Best regards

**Comment by coskrrb2002** [comment, marked as answer]
I guess I have to use UserObject... Instead of auxkernel...

**Reply by coskrrb2002** [reply]
No, I got it. I am using materials, and it works!

**Reply by ngrilli** [reply]
Dear @coskrrb2002 

What I usually do is the following:
If I need a quantity to use in a kernel, I define it as a material property in a material object.
Then the quantity can be imported in a kernel using getMaterialProperty.

I use auxkernels to calculate auxvariables that are needed into material objects
or to output material properties for visualisation.

Best Regards,
Nicolò Grilli

**Reply by coskrrb2002** [reply]
Dear @ngrilli ,

Thank you for your advice!


Best regards,
Nakkyu
